### PR TITLE
Add compositional heating model

### PR DESCRIPTION
**Pull request by naliboff**
Lithospheric-scale models often assign fixed heat generation (W/m^3) values to distinct materials. This pull request provides this functionality by adding a heating model where one may assign temporally constant values of heat production to compositional fields.

### CODE REVIEW

File: include/aspect/heating_model/compositional_heating.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        std::vector<double> compute_volume_fractions(
+          const std::vector<double> &compositional_fields) const;
+
+        /**
+         * Magnitude of heat production in each compositional field
+         */
+        std::vector<double> heating_values;
+
+        /**
+         * User can choose whether compositional field is used in heat production averaging
```
Comment (gassmoeller) [inline include/aspect/heating_model/compositional_heating.h]: Also would it make sense to use a `std::vector<bool>` ?

File: source/heating_model/compositional_heating.cc [hunk 2]
```diff
@@ -0,0 +1,172 @@
+/*
+  Copyright (C) 2011 - 2017 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline source/heating_model/compositional_heating.cc]: 2017

File: source/heating_model/compositional_heating.cc [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+
+#include <aspect/heating_model/compositional_heating.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
```
Comment (gassmoeller) [inline source/heating_model/compositional_heating.cc]: do you need the geometry model interface header?

File: source/heating_model/compositional_heating.cc [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        prm.enter_subsection("Compositional heating");
+        {
+          prm.declare_entry("Compositional heating values","0",
+                            Patterns::List (Patterns::Double(0)),
+                            "List of heat production per unit volume values for "
+                            "background and compositional fields, for a total of "
+                            "N+1 values, where N is the number of compositional fields. "
+                            "Units: W/m3.");
+          prm.declare_entry ("Use compositional field for heat production averaging", "1",
+                             Patterns::List(Patterns::Integer(0,1)),
```
Comment (gassmoeller) [inline source/heating_model/compositional_heating.cc]: Would a `Patterns::List(Patterns::Bool)` work? Then the input would look like `true,false,false`. If you prefer 0 and 1 in the input file keeping the integer here would be fine with me as well, I would just like to store it internally as bool to make its purpose clear.

File: source/heating_model/compositional_heating.cc [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+              x_comp[i] = std::min(std::max(x_comp[i], 0.0), 1.0);
+              sum_composition += x_comp[i];
+            }
+          else
+            x_comp[i] = 0.0;
+        }
+
+      if (sum_composition >= 1.0)
+        {
+          volume_fractions[0] = 0.0;  //background material
```
Comment (gassmoeller) [inline source/heating_model/compositional_heating.cc]: So this means if the compositions happen to add up to more than 1 you assume there is no background material? How does that work with the input file parameters? You are still expected to input a background value? This feels strange. 

Ah, but I just checked the multicomponent material model and that one works the same. So lets not worry about it for now. I would prefer to have it more explicit, whether somebody uses 1 value (expanded to all fields), N values (one per field) or N+1 values (one per field plus background field), but we should tackle that as a separate project. I opened an issue about this.

File: source/heating_model/compositional_heating.cc [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+      for (unsigned int q = 0; q < heating_model_outputs.heating_source_terms.size(); ++q)
+        {
+
+          // Compute compositional volume fractions
+          const std::vector<double> volume_fractions = compute_volume_fractions(material_model_inputs.composition[q]);
+
+          // Calculate average compositional heat production
+          double compositional_heat_production = 0.;
+          for (unsigned int c=0; c< volume_fractions.size(); ++c)
```
Comment (gassmoeller) [inline source/heating_model/compositional_heating.cc]: missing space in the comparison

File: include/aspect/heating_model/compositional_heating.h [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        /**
+         * Read the parameters this class declares from the parameter file.
+         */
+        virtual
+        void
+        parse_parameters (ParameterHandler &prm);
+
+      private:
+
+        std::vector<double> compute_volume_fractions(
```
Comment (naliboff) [inline include/aspect/heating_model/compositional_heating.h]: I added in the description from the multicomponent material model. Given it makes some fairly significant assumptions (i.e. your comment below) probably best to make sure it is clearly explained everywhere.

File: source/heating_model/compositional_heating.cc (hunk 4 above)
Comment (naliboff) [inline source/heating_model/compositional_heating.cc]: I've addressed all comments except this one. I tried changing this parameter to a bool list and it was not able to compile successfully. I looked through ASPECT and did not have any other instances of bool list. Is it in fact possible to do this?

**Comment by gassmoeller** [comment]
Ok, thanks. Lets not worry about the bool list then.

**Pull request merged**